### ISSUE TITLE
NamedEntity Recognizer doesn't train well for Chinese

### ISSUE THREAD

**Issue by lingvisa** [explosion/spaCy#5112]
I modified this English NE tagging example slightly with my own Chinese example data and the blank Chinese model: https://spacy.io/usage/examples. This English example works well.

My Chinese version of this example is almost exactly the same as English and looks like below:
```
# training data
TRAIN_DATA = [
    ("谁是毛泽东？", {"entities": [(2, 4, "PERSON")]}),
    ("他是毛泽东？", {"entities": [(2, 4, "PERSON")]}),
    ("他是习近平？", {"entities": [(2, 4, "PERSON")]}),
    ("他是习近平吗？", {"entities": [(2, 4, "PERSON")]}),
    ("他不是毛泽东？", {"entities": [(3, 5, "PERSON")]}),
    ("他是不是毛泽东？", {"entities": [(4, 6, "PERSON")]}),
    ("他还是毛泽东？", {"entities": [(3, 5, "PERSON")]}),
    ("他是毛泽东吧？", {"entities": [(2, 4, "PERSON")]}),
    ("我喜欢北京和武汉", {"entities": [(3, 4, "LOC"), (6, 7, "LOC")]}),
    ("我喜欢北京和天津的食物", {"entities": [(3, 4, "LOC"), (6, 7, "LOC")]}),
    ("我喜欢北京和武汉吧", {"entities": [(3, 4, "LOC"), (6, 7, "LOC")]}),
    ("我喜欢北京", {"entities": [(3, 4, "LOC")]}),
]

@plac.annotations(
    model=("Model name. Defaults to blank 'en' model.", "option", "m", str),
    output_dir=("Optional output directory", "option", "o", Path),
    n_iter=("Number of training iterations", "option", "n", int),
)
def main(model=None, output_dir=None, n_iter=300):
    """Load the model, set up the pipeline and train the entity recognizer."""
    if model is not None:
        nlp = spacy.load(model)  # load existing spaCy model
        print("Loaded model '%s'" % model)
    else:
        nlp = spacy.blank("zh")  # create blank Language class
        print("Created blank 'zh' model")

    # create the built-in pipeline components and add them to the pipeline
    # nlp.create_pipe works for built-ins that are registered with spaCy    if "ner" not in nlp.pipe_names:
    if "ner" not in nlp.pipe_names:
        ner = nlp.create_pipe("ner")
        nlp.add_pipe(ner, last=True)
    # otherwise, get it so we can add labels
    else:
        ner = nlp.get_pipe("ner")

    # add labels
    for _, annotations in TRAIN_DATA:
        for ent in annotations.get("entities"):
            ner.add_label(ent[2])

    # get names of other pipes to disable them during training
    pipe_exceptions = ["ner", "trf_wordpiecer", "trf_tok2vec"]
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe not in pipe_exceptions]
    with nlp.disable_pipes(*other_pipes):  # only train NER
        # reset and initialize the weights randomly – but only if we're
        # training a new model
        if model is None:
            nlp.begin_training()
        for itn in range(n_iter):
            random.shuffle(TRAIN_DATA)
            losses = {}
            # batch up the examples using spaCy's minibatch
            batches = minibatch(TRAIN_DATA, size=compounding(4.0, 32.0, 1.001))
            for batch in batches:
                texts, annotations = zip(*batch)
                nlp.update(
                    texts,  # batch of texts
                    annotations,  # batch of annotations
                    drop=0.5,  # dropout - make it harder to memorise data
                    losses=losses,
                )
            print("Losses", losses)

    # test the trained model
    for text, _ in TRAIN_DATA:
        doc = nlp(text)
        print("Entities", [(ent.text, ent.label_) for ent in doc.ents])
        print("Tokens", [(t.text, t.ent_type_, t.ent_iob) for t in doc])

    # save model to output directory
    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()
        nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the saved model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        for text, _ in TRAIN_DATA:
            doc = nlp2(text)
            print("Entities", [(ent.text, ent.label_) for ent in doc.ents])
            print("Tokens", [(t.text, t.ent_type_, t.ent_iob) for t in doc])
if __name__ == "__main__":
    plac.call(main)

```

However, this doesn't train anything. The loss remains high at the end of the iterations:
```
Losses {'ner': 7.137901513094454e-16}
Losses {'ner': 2.679321975462729e-18}
Losses {'ner': 7.589459987809973e-18}
Losses {'ner': 1.0718421745284892e-16}
Losses {'ner': 5.812764809854231e-17}
Losses {'ner': 5.848371167905202e-19}
Losses {'ner': 3.327016094287402e-17}
Losses {'ner': 5.769298805678327e-18}
Entities []
Tokens [('他', '', 2), ('是', '', 2), ('习近平', '', 2), ('吗', '', 2)]
Entities []
Tokens [('他', '', 2), ('是', '', 2), ('毛泽东', '', 2)]
Entities []
Tokens [('他', '', 2), ('还是', '', 2), ('毛泽东', '', 2)]
Entities []
Tokens [('他', '', 2), ('是', '', 2), ('习近平', '', 2)]
Entities []
Tokens [('我', '', 2), ('喜欢', '', 2), ('北京', '', 2), ('和', '', 2), ('武汉', '', 2)]
Entities []
Tokens [('我', '', 2), ('喜欢', '', 2), ('北京', '', 2)]
Entities []
Tokens [('他', '', 2), ('是不是', '', 2), ('毛泽东', '', 2)]
Entities []
Tokens [('我', '', 2), ('喜欢', '', 2), ('北京', '', 2), ('和', '', 2), ('天津', '', 2), ('的', '', 2), ('食物', '', 2)]


 
```

Do I need to do some special configuration in the Chinese Jieba tokenizer?

**Comment by adrianeboyd** [comment]
I think the main problem is that the annotation offsets in `TRAIN_DATA` are incorrect. The character offsets should be the same as string slices (the end character is exclusive). For the first text, the annotation says that `毛泽` is the `PERSON`.

```
assert "谁是毛泽东？"[2:4] == "毛泽"
```

You can check whether the tokenization lines up with the character offsets like this:

```
import spacy
from spacy.gold import biluo_tags_from_offsets
from spacy.lang.zh import Chinese

for text, annot in TRAIN_DATA:
    doc = nlp.make_doc(text)
    biluo = biluo_tags_from_offsets(doc, annot["entities"])
    print([t.text for t in doc], biluo)
```

You want to see `B/I/L/U-PERSON` tags in this output, `-` means the annotation couldn't be aligned with the tokenization, so the NER model doesn't learn anything from this token while training:

```
['谁', '是', '毛泽东', '？'] ['O', 'O', '-', 'O']
['他', '是', '毛泽东', '？'] ['O', 'O', '-', 'O']
['他', '是', '习近平', '？'] ['O', 'O', '-', 'O']
['他', '是', '习近平', '吗', '？'] ['O', 'O', '-', 'O', 'O']
['他', '不是', '毛泽东', '？'] ['O', 'O', '-', 'O']
['他', '是不是', '毛泽东', '？'] ['O', 'O', '-', 'O']
['他', '还是', '毛泽东', '？'] ['O', 'O', '-', 'O']
['他', '是', '毛泽东', '吧', '？'] ['O', 'O', '-', 'O', 'O']
['我', '喜欢', '北京', '和', '武汉'] ['O', 'O', '-', 'O', '-']
['我', '喜欢', '北京', '和', '天津', '的', '食物'] ['O', 'O', '-', 'O', '-', 'O', 'O']
['我', '喜欢', '北京', '和', '武汉', '吧'] ['O', 'O', '-', 'O', '-', 'O']
['我', '喜欢', '北京'] ['O', 'O', '-']
```

Other notes:

* the loss is extremely low at the end, since it only sees `O` and has learned that nothing is an entity
* your example output is not from the same training data as shown in the script?

**Comment by lingvisa** [comment]
Yes. The offset is the problem. Thanks for pointing this out and the code for validation.
